File: non-kernel/frontend/scripts/verify-art-l7-replay-001.mjs
#!/usr/bin/env node
/**
 * Independent reproduction for ART-L7-REPLAY-001.
 * Does not import the website UI or trust its claims.
 *
 * Usage:
 *   node non-kernel/frontend/scripts/verify-art-l7-replay-001.mjs
 *   node non-kernel/frontend/scripts/verify-art-l7-replay-001.mjs path/to/ART-L7-REPLAY-001.json
 *
 * Exit 0 = sealed hashes reproduced independently.
 * Exit 1 = mismatch or error.
 */

import { createHash } from "crypto";
import { readFileSync } from "fs";
import { dirname, join } from "path";
import { fileURLToPath } from "url";

const ALLOWED = {
  INGESTED: new Set(["NORMALIZED"]),
  NORMALIZED: new Set(["VERIFIED"]),
  VERIFIED: new Set(["CORRELATED"]),
  CORRELATED: new Set(["ARCHIVED"]),
  ARCHIVED: new Set(),
  CONTESTED: new Set(),
};

function canonicalize(value) {
  if (value === null || typeof value !== "object") return JSON.stringify(value);
  if (Array.isArray(value)) return `[${value.map(canonicalize).join(",")}]`;
  const keys = Object.keys(value).sort();
  return `{${keys.map((k) => `${JSON.stringify(k)}:${canonicalize(value[k])}`).join(",")}}`;
}

function sha256Hex(input) {
  return createHash("sha256").update(input, "utf8").digest("hex");
}

function reduceEvent(state, event) {
  const next = {
    lifecycle_state: state.lifecycle_state,
    records: { ...state.records },
    commit_finalized: state.commit_finalized,
  };
  if (event.event_type === "RECORD_INSERT") {
    if (next.records[event.record_id]) throw new Error("duplicate record_id");
    next.records[event.record_id] = event.payload;
  } else if (event.event_type === "STATE_TRANSITION") {
    const st = next.lifecycle_state;
    const t = event.to_state;
    if (!(t === "CONTESTED" || ALLOWED[st]?.has(t))) {
      throw new Error(`illegal transition ${st}->${t}`);
    }
    next.lifecycle_state = t;
  } else if (event.event_type === "COMMIT_FINALIZED") {
    next.commit_finalized = true;
  } else {
    throw new Error(`unsupported event_type ${event.event_type}`);
  }
  return next;
}

function replay(events, initial) {
  let state = structuredClone(initial);
  for (const event of events) state = reduceEvent(state, event);
  return state;
}

function ledgerHead(events) {
  let head = "GENESIS";
  for (const event of events) {
    head = sha256Hex(`${head}|${canonicalize(event)}`);
  }
  return head;
}

function verify(capsule) {
  const runA = replay(capsule.events, capsule.initial_state);
  const runB = replay(capsule.events, capsule.initial_state);
  const stateHashA = sha256Hex(canonicalize(runA));
  const stateHashB = sha256Hex(canonicalize(runB));
  const ledger_head_hash = ledgerHead(capsule.events);
  const receipt_payload = {
    version_bundle: capsule.version_bundle,
    event_count: capsule.events.length,
    state_hash: stateHashA,
    ledger_head_hash,
    terminal_lifecycle: runA.lifecycle_state,
    commit_finalized: runA.commit_finalized,
  };
  const receipt_hash = sha256Hex(canonicalize(receipt_payload));

  const checks = {
    independent_parity: stateHashA === stateHashB,
    state_hash: stateHashA === capsule.expected.state_hash,
    ledger_head_hash: ledger_head_hash === capsule.expected.ledger_head_hash,
    receipt_hash: receipt_hash === capsule.expected.receipt_hash,
    terminal_lifecycle: runA.lifecycle_state === capsule.expected.terminal_lifecycle,
    commit_finalized: runA.commit_finalized === capsule.expected.commit_finalized,
  };

  return {
    artifactId: capsule.artifactId,
    pass: Object.values(checks).every(Boolean),
    checks,
    computed: {
      state_hash: stateHashA,
      receipt_hash,
      ledger_head_hash,
    },
    expected: capsule.expected,
  };
}

const here = dirname(fileURLToPath(import.meta.url));
const defaultPath = join(here, "../data/evidence/artifacts/ART-L7-REPLAY-001.json");
const path = process.argv[2] || defaultPath;

let capsule;
try {
  capsule = JSON.parse(readFileSync(path, "utf8"));
} catch (err) {
  console.error("FAIL: cannot read capsule", path, err.message);
  process.exit(1);
}

if (capsule.artifactId !== "ART-L7-REPLAY-001") {
  console.error("FAIL: unexpected artifactId", capsule.artifactId);
  process.exit(1);
}

const result = verify(capsule);
console.log(JSON.stringify(result, null, 2));
process.exit(result.pass ? 0 : 1);

File: non-kernel/frontend/scripts/verify_art_l7_replay_001.py
#!/usr/bin/env python3
"""Independent Python reproduction for ART-L7-REPLAY-001.

No website UI. No Node dependency. Same sealed capsule as the JS verifier.
Exit 0 = match sealed expected hashes.
"""
from __future__ import annotations

import hashlib
import json
import sys
from copy import deepcopy
from pathlib import Path

ALLOWED = {
    "INGESTED": {"NORMALIZED"},
    "NORMALIZED": {"VERIFIED"},
    "VERIFIED": {"CORRELATED"},
    "CORRELATED": {"ARCHIVED"},
    "ARCHIVED": set(),
    "CONTESTED": set(),
}


def canonicalize(value) -> str:
    if value is None or not isinstance(value, (dict, list)):
        return json.dumps(value, separators=(",", ":"), ensure_ascii=False)
    if isinstance(value, list):
        return "[" + ",".join(canonicalize(v) for v in value) + "]"
    keys = sorted(value.keys())
    return "{" + ",".join(json.dumps(k, ensure_ascii=False) + ":" + canonicalize(value[k]) for k in keys) + "}"


def sha256_hex(s: str) -> str:
    return hashlib.sha256(s.encode("utf-8")).hexdigest()


def reduce_event(state: dict, event: dict) -> dict:
    next_state = {
        "lifecycle_state": state["lifecycle_state"],
        "records": dict(state["records"]),
        "commit_finalized": state["commit_finalized"],
    }
    et = event["event_type"]
    if et == "RECORD_INSERT":
        rid = event["record_id"]
        if rid in next_state["records"]:
            raise ValueError("duplicate record_id")
        next_state["records"][rid] = event["payload"]
    elif et == "STATE_TRANSITION":
        st = next_state["lifecycle_state"]
        t = event["to_state"]
        if not (t == "CONTESTED" or t in ALLOWED.get(st, set())):
            raise ValueError(f"illegal transition {st}->{t}")
        next_state["lifecycle_state"] = t
    elif et == "COMMIT_FINALIZED":
        next_state["commit_finalized"] = True
    else:
        raise ValueError(f"unsupported event_type {et}")
    return next_state


def replay(events, initial):
    state = deepcopy(initial)
    for event in events:
        state = reduce_event(state, event)
    return state


def ledger_head(events):
    head = "GENESIS"
    for event in events:
        head = sha256_hex(f"{head}|{canonicalize(event)}")
    return head


def verify(capsule: dict) -> dict:
    run_a = replay(capsule["events"], capsule["initial_state"])
    run_b = replay(capsule["events"], capsule["initial_state"])
    state_hash_a = sha256_hex(canonicalize(run_a))
    state_hash_b = sha256_hex(canonicalize(run_b))
    head = ledger_head(capsule["events"])
    receipt_payload = {
        "version_bundle": capsule["version_bundle"],
        "event_count": len(capsule["events"]),
        "state_hash": state_hash_a,
        "ledger_head_hash": head,
        "terminal_lifecycle": run_a["lifecycle_state"],
        "commit_finalized": run_a["commit_finalized"],
    }
    receipt_hash = sha256_hex(canonicalize(receipt_payload))
    exp = capsule["expected"]
    checks = {
        "independent_parity": state_hash_a == state_hash_b,
        "state_hash": state_hash_a == exp["state_hash"],
        "ledger_head_hash": head == exp["ledger_head_hash"],
        "receipt_hash": receipt_hash == exp["receipt_hash"],
        "terminal_lifecycle": run_a["lifecycle_state"] == exp["terminal_lifecycle"],
        "commit_finalized": run_a["commit_finalized"] == exp["commit_finalized"],
    }
    return {
        "implementation": "python3",
        "artifactId": capsule["artifactId"],
        "pass": all(checks.values()),
        "checks": checks,
        "computed": {
            "state_hash": state_hash_a,
            "receipt_hash": receipt_hash,
            "ledger_head_hash": head,
        },
        "expected": exp,
    }


def main() -> int:
    default = Path(__file__).resolve().parent.parent / "data/evidence/artifacts/ART-L7-REPLAY-001.json"
    path = Path(sys.argv[1]) if len(sys.argv) > 1 else default
    capsule = json.loads(path.read_text(encoding="utf-8"))
    if capsule.get("artifactId") != "ART-L7-REPLAY-001":
        print(json.dumps({"pass": False, "error": "unexpected artifactId"}))
        return 1
    result = verify(capsule)
    print(json.dumps(result, indent=2))
    return 0 if result["pass"] else 1


if __name__ == "__main__":
    raise SystemExit(main())
